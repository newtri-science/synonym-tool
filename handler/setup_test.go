package handler_test

import (
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"testing"

	"github.com/labstack/echo/v4"
	"github.com/stretchr/testify/assert"
	"go.uber.org/mock/gomock"
	"go.uber.org/zap/zaptest"

	"github.com/michelm117/cycling-coach-lab/handler"
	"github.com/michelm117/cycling-coach-lab/mocks"
	"github.com/michelm117/cycling-coach-lab/model"
	"github.com/michelm117/cycling-coach-lab/test_utils"
)

func TestRenderSetup(t *testing.T) {
	logger := zaptest.NewLogger(t).Sugar()

	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mg := mocks.NewMockGlobalSettingServicer(ctrl)

	t.Run("App already initialized", func(t *testing.T) {
		mg.EXPECT().IsAppInitialized().Return(true)

		handler := handler.NewSetupHandler(mg, nil, nil, logger)

		// Create a request
		req := httptest.NewRequest(http.MethodGet, "/setup", nil)
		rec := httptest.NewRecorder()
		c := echo.New().NewContext(req, rec)

		// Call the handler
		assert.NoError(t, handler.RenderSetup(c))
		assert.Equal(t, http.StatusTemporaryRedirect, rec.Code)
	})

	t.Run("App not initialized", func(t *testing.T) {
		mg.EXPECT().IsAppInitialized().Return(false)

		handler := handler.NewSetupHandler(mg, nil, nil, logger)

		// Create a request
		req := httptest.NewRequest(http.MethodGet, "/setup", nil)
		rec := httptest.NewRecorder()
		c := echo.New().NewContext(req, rec)

		// Call the handler
		assert.NoError(t, handler.RenderSetup(c))
		assert.Equal(t, http.StatusOK, rec.Code)
		test_utils.MakeSnapshot(t, rec.Body.String())
	})
}

func createSetupRequest(firstname, lastname, email, password, confirmPassword, dateOfBirth string) *http.Request {
	form := url.Values{}
	form.Add("firstname", firstname)
	form.Add("lastname", lastname)
	form.Add("email", email)
	form.Add("password", password)
	form.Add("confirmPassword", confirmPassword)
	form.Add("dateOfBirth", dateOfBirth)

	req := httptest.NewRequest(http.MethodPost, "/setup", strings.NewReader(form.Encode()))
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	return req
}

func TestSetup(t *testing.T) {
	logger := zaptest.NewLogger(t).Sugar()
	ctrl := gomock.NewController(t)
	mg := mocks.NewMockGlobalSettingServicer(ctrl)
	mv := mocks.NewMockValidator(ctrl)
	mu := mocks.NewMockUserServicer(ctrl)

	t.Run("App already initialized", func(t *testing.T) {
		mg.EXPECT().IsAppInitialized().Return(true)

		handler := handler.NewSetupHandler(mg, nil, nil, logger)

		// Create a request
		req := createSetupRequest("John", "Doe", "john@doe.com", "password", "password", "1990-01-01")
		rec := httptest.NewRecorder()
		c := echo.New().NewContext(req, rec)

		// Call the handler
		assert.ErrorContains(t, handler.Setup(c), "App already initialized")
		assert.Equal(t, http.StatusOK, rec.Code)
	})

	t.Run("Invalid input", func(t *testing.T) {
		mg.EXPECT().IsAppInitialized().Return(false)

		mv.EXPECT().CreateValidUser(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, assert.AnError)
		handler := handler.NewSetupHandler(mg, nil, mv, logger)

		// Create a request
		req := createSetupRequest("John", "Doe", "john@doe.com", "password", "password", "invalid")
		rec := httptest.NewRecorder()
		c := echo.New().NewContext(req, rec)

		// Call the handler
		assert.Error(t, handler.Setup(c))
		assert.Equal(t, http.StatusOK, rec.Code)
	})

	t.Run("Success", func(t *testing.T) {
		user := &model.User{
			ID:        1,
			Firstname: "John",
			Lastname:  "Doe",
			Email:     "john@doe.com",
		}
		mg.EXPECT().IsAppInitialized().Return(false)
		mg.EXPECT().InitializeApp().Return(nil)
		mv.EXPECT().CreateValidUser(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(user, nil)
		mu.EXPECT().AddUser(gomock.Any()).Return(nil, nil)

		handler := handler.NewSetupHandler(mg, mu, mv, logger)

		// Create a request
		req := createSetupRequest(user.Firstname, user.Lastname, user.Email, "password", "password", "1990-01-01")
		rec := httptest.NewRecorder()
		c := echo.New().NewContext(req, rec)

		// Call the handler
		assert.NoError(t, handler.Setup(c))
		assert.Equal(t, http.StatusOK, rec.Code)
	})
}
