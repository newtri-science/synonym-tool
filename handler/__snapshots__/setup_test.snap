
[TestRenderSetup/App_not_initialized - 1]
<!doctype html>
<html data-theme="dark">
  <head>
    <title>
      Cycling Coach Lab
    </title>
    <meta charset="utf-8">
    <meta name="robots" content="noindex">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, interactive-widget=resizes-content">
    <meta name="htmx-config" content="{&#34;useTemplateFragments&#34;:true}">
    <link rel="icon" href="data:image/svg+xml,&lt;svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 100 100%22&gt;&lt;text y=%22.9em%22 font-size=%2290%22&gt;ðŸš´&lt;/text&gt;&lt;/svg&gt;">
    <link rel="stylesheet" href="/assets/styles.css">
    <link rel="stylesheet" href="/assets/base.css">
    <!-- htmx -->
    <script src="https://unpkg.com/htmx.org@1.9.10" integrity="sha384-D1Kt99CQMDuVetoL1lrYwg5t+9QdHe7NLX/SoJYkXDFfX37iInKRy5xLSi8nO7UC" crossorigin="anonymous"></script>
    <script src="https://unpkg.com/htmx.org/dist/ext/path-params.js"></script>
    <script src="https://unpkg.com/htmx.org/dist/ext/response-targets.js"></script>
    <!-- Alpine Plugins -->
    <script defer src="https://cdn.jsdelivr.net/npm/@alpinejs/persist@3.x.x/dist/cdn.min.js"></script>
    <script defer src="https://cdn.jsdelivr.net/npm/@alpinejs/focus@3.x.x/dist/cdn.min.js"></script>
    <!-- Alpine Core -->
    <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
    <!-- Font Awesome -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.2/css/fontawesome.min.css" integrity="sha512-P9vJUXK+LyvAzj8otTOKzdfF1F3UYVl13+F8Fof8/2QNb8Twd6Vb+VD52I7+87tex9UXxnzPgWA3rH96RExA7A==" crossorigin="anonymous" referrerpolicy="no-referrer">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.2/css/solid.min.css" integrity="sha512-tk4nGrLxft4l30r9ETuejLU0a3d7LwMzj0eXjzc16JQj+5U1IeVoCuGLObRDc3+eQMUcEQY1RIDPGvuA7SNQ2w==" crossorigin="anonymous" referrerpolicy="no-referrer">
    <script src="https://kit.fontawesome.com/e219cbc312.js" crossorigin="anonymous"></script>
    <script src="https://cdn.jsdelivr.net/npm/sweetalert2@11"></script>
  </head>
  <body class="font-sans" x-init="
            window.toast = (message, options = {}) =&gt; {
                let description = &#39;&#39;;
                let type = &#39;default&#39;;
                let position = &#39;top-right&#39;;
                if(typeof options.description != &#39;undefined&#39;) description = options.description;
                if(typeof options.type != &#39;undefined&#39;) type = options.type;
                if(typeof options.position != &#39;undefined&#39;) position = options.position;
                
                window.dispatchEvent(new CustomEvent(&#39;toast-show&#39;, { 
                    detail : { 
                      type: type, 
                      message: message, 
                      description: description, 
                      position : position 
                    }
                  }
                ));
            };
        ">
    <div class="flex items-center justify-center h-full">
      <div class="rounded bg-base-200 shadow-lg p-8 max-w-md w-full">
        <form id="signupForm" hx-post="/setup" class="space-y-4">
          <h2 class="text-3xl font-bold text-center">
            Cycling Coach Lab
          </h2>
          <h2 class="text-2xl font-bold text-center">
            Application Setup
          </h2>
          <p class="text-center">
            FÃ¼lle die folgenden Felder aus, um die Anwendung zu konfigurieren und deinen Account zu erstellen.
          </p>
          <div class="mb-4">
            <label class="block text-sm font-semibold text-gray-600">
              Vorname
            </label>
            <input class="w-full input input-bordered" name="firstname" type="text" placeholder="Vorname">
          </div>
          <div class="mb-4">
            <label class="block text-sm font-semibold text-gray-600">
              Name
            </label>
            <input class="w-full input input-bordered" name="lastname" type="text" placeholder="Name">
          </div>
          <div class="mb-4">
            <label class="block text-sm font-semibold text-gray-600">
              Email
            </label>
            <input class="w-full input input-bordered" name="email" type="email" placeholder="Email">
          </div>
          <div class="mb-4">
            <label class="block text-sm font-semibold text-gray-600">
              Passwort
            </label>
            <input class="w-full input input-bordered" name="password" type="password" placeholder="Passwort">
          </div>
          <div class="mb-4">
            <label class="block text-sm font-semibold text-gray-600">
              Passwort wiederholen
            </label>
            <input class="w-full input input-bordered" name="confirmPassword" type="password" placeholder="Passwort wiederholen">
          </div>
          <div class="flex flex-col">
            <label for="dateOfBirth" class="block text-sm font-semibold text-gray-600">
              Geburtsdatum:
            </label>
            <input class="block text-sm font-semibold text-gray-600" type="date" id="dateOfBirth" name="dateOfBirth" required>
          </div>
          <div class="flex items-center justify-between">
            <button class="btn btn-primary w-full" type="submit">
              Registrieren
            </button>
          </div>
        </form>
      </div>
    </div>
    <template x-teleport="body">
      <ul x-data="{ 
                toasts: [],
                toastsHovered: false,
                expanded: false,
                layout: &#39;default&#39;,
                position: &#39;top-center&#39;,
                paddingBetweenToasts: 15,
                deleteToastWithId (id){
                    for(let i = 0; i &lt; this.toasts.length; i++){
                        if(this.toasts[i].id === id){
                            this.toasts.splice(i, 1);
                            break;
                        }
                    }
                },
                burnToast(id){
                    burnToast = this.getToastWithId(id);
                    burnToastElement = document.getElementById(burnToast.id);
                    if(burnToastElement){
                        if(this.toasts.length == 1){
                            if(this.layout==&#39;default&#39;){
                                this.expanded = false;
                            }
                            burnToastElement.classList.remove(&#39;translate-y-0&#39;);
                            if(this.position.includes(&#39;bottom&#39;)){
                                burnToastElement.classList.add(&#39;translate-y-full&#39;);
                            } else {
                                burnToastElement.classList.add(&#39;-translate-y-full&#39;);
                            }
                            burnToastElement.classList.add(&#39;-translate-y-full&#39;);
                        }
                        burnToastElement.classList.add(&#39;opacity-0&#39;);
                        let that = this;
                        setTimeout(function(){
                            that.deleteToastWithId(id);
                            setTimeout(function(){
                                that.stackToasts();
                            }, 1)
                        }, 300);
                    }
                },
                getToastWithId(id){
                    for(let i = 0; i &lt; this.toasts.length; i++){
                        if(this.toasts[i].id === id){
                            return this.toasts[i];
                        }
                    }
                },
                stackToasts(){
                    this.positionToasts();
                    this.calculateHeightOfToastsContainer();
                    let that = this;
                    setTimeout(function(){
                        that.calculateHeightOfToastsContainer();
                    }, 300);
                },
                positionToasts(){
                    if(this.toasts.length == 0) return;
                    let topToast = document.getElementById( this.toasts[0].id );
                    topToast.style.zIndex = 100;
                    if(this.expanded){
                        if(this.position.includes(&#39;bottom&#39;)){
                            topToast.style.top = &#39;auto&#39;;
                            topToast.style.bottom = &#39;0px&#39;;
                        } else {
                            topToast.style.top = &#39;0px&#39;;
                        }
                    }

                    let bottomPositionOfFirstToast = this.getBottomPositionOfElement(topToast);

                    if(this.toasts.length == 1) return;
                    let middleToast = document.getElementById( this.toasts[1].id );
                    middleToast.style.zIndex = 90;

                    if(this.expanded){
                        middleToastPosition = topToast.getBoundingClientRect().height +
                                                this.paddingBetweenToasts + &#39;px&#39;;

                        if(this.position.includes(&#39;bottom&#39;)){
                            middleToast.style.top = &#39;auto&#39;;
                            middleToast.style.bottom = middleToastPosition;
                        } else {
                            middleToast.style.top = middleToastPosition;
                        }

                        middleToast.style.scale = &#39;100%&#39;;
                        middleToast.style.transform = &#39;translateY(0px)&#39;;
                        
                    } else {
                        middleToast.style.scale = &#39;94%&#39;;
                        if(this.position.includes(&#39;bottom&#39;)){
                            middleToast.style.transform = &#39;translateY(-16px)&#39;;
                        } else {
                            this.alignBottom(topToast, middleToast);
                            middleToast.style.transform = &#39;translateY(16px)&#39;;
                        }
                    }
                    

                    if(this.toasts.length == 2) return;
                    let bottomToast = document.getElementById( this.toasts[2].id );
                    bottomToast.style.zIndex = 80;
                    if(this.expanded){
                        bottomToastPosition = topToast.getBoundingClientRect().height + 
                                                this.paddingBetweenToasts + 
                                                middleToast.getBoundingClientRect().height +
                                                this.paddingBetweenToasts + &#39;px&#39;;
                        
                        if(this.position.includes(&#39;bottom&#39;)){
                            bottomToast.style.top = &#39;auto&#39;;
                            bottomToast.style.bottom = bottomToastPosition;
                        } else {
                            bottomToast.style.top = bottomToastPosition;
                        }

                        bottomToast.style.scale = &#39;100%&#39;;
                        bottomToast.style.transform = &#39;translateY(0px)&#39;;
                    } else {
                        bottomToast.style.scale = &#39;88%&#39;;
                        if(this.position.includes(&#39;bottom&#39;)){
                            bottomToast.style.transform = &#39;translateY(-32px)&#39;;
                        } else {
                            this.alignBottom(topToast, bottomToast);
                            bottomToast.style.transform = &#39;translateY(32px)&#39;;
                        }
                    }

                    

                    if(this.toasts.length == 3) return;
                    let burnToast = document.getElementById( this.toasts[3].id );
                    burnToast.style.zIndex = 70;
                    if(this.expanded){
                        burnToastPosition = topToast.getBoundingClientRect().height + 
                                                this.paddingBetweenToasts + 
                                                middleToast.getBoundingClientRect().height + 
                                                this.paddingBetweenToasts + 
                                                bottomToast.getBoundingClientRect().height + 
                                                this.paddingBetweenToasts + &#39;px&#39;;
                        
                        if(this.position.includes(&#39;bottom&#39;)){
                            burnToast.style.top = &#39;auto&#39;;
                            burnToast.style.bottom = burnToastPosition;
                        } else {
                            burnToast.style.top = burnToastPosition;
                        }

                        burnToast.style.scale = &#39;100%&#39;;
                        burnToast.style.transform = &#39;translateY(0px)&#39;;
                    } else {
                        burnToast.style.scale = &#39;82%&#39;;
                        this.alignBottom(topToast, burnToast);
                        burnToast.style.transform = &#39;translateY(48px)&#39;;
                    }

                    burnToast.firstElementChild.classList.remove(&#39;opacity-100&#39;);
                    burnToast.firstElementChild.classList.add(&#39;opacity-0&#39;);

                    let that = this;
                    // Burn ðŸ”¥ (remove) last toast
                    setTimeout(function(){
                            that.toasts.pop();
                        }, 300);

                    if(this.position.includes(&#39;bottom&#39;)){
                            middleToast.style.top = &#39;auto&#39;;
                    }

                    return;
                },
                alignBottom(element1, element2) {
                    // Get the top position and height of the first element
                    let top1 = element1.offsetTop;
                    let height1 = element1.offsetHeight;

                    // Get the height of the second element
                    let height2 = element2.offsetHeight;

                    // Calculate the top position for the second element
                    let top2 = top1 + (height1 - height2);

                    // Apply the calculated top position to the second element
                    element2.style.top = top2 + &#39;px&#39;;
                },
                alignTop(element1, element2) {
                    // Get the top position of the first element
                    let top1 = element1.offsetTop;

                    // Apply the same top position to the second element
                    element2.style.top = top1 + &#39;px&#39;;
                },
                resetBottom(){
                    for(let i = 0; i &lt; this.toasts.length; i++){
                        if(document.getElementById( this.toasts[i].id )){
                            let toastElement = document.getElementById( this.toasts[i].id );
                            toastElement.style.bottom = &#39;0px&#39;;
                        }
                    }
                },
                resetTop(){
                    for(let i = 0; i &lt; this.toasts.length; i++){
                        if(document.getElementById( this.toasts[i].id )){
                            let toastElement = document.getElementById( this.toasts[i].id );
                            toastElement.style.top = &#39;0px&#39;;
                        }
                    }
                },
                getBottomPositionOfElement(el){
                    return (el.getBoundingClientRect().height + el.getBoundingClientRect().top);
                },
                calculateHeightOfToastsContainer(){
                    if(this.toasts.length == 0){
                        $el.style.height = &#39;0px&#39;;
                        return;
                    }

                    lastToast = this.toasts[this.toasts.length - 1];
                    lastToastRectangle = document.getElementById(lastToast.id).getBoundingClientRect();
                    
                    firstToast = this.toasts[0];
                    firstToastRectangle = document.getElementById(firstToast.id).getBoundingClientRect();

                    if(this.toastsHovered){
                        if(this.position.includes(&#39;bottom&#39;)){
                            $el.style.height = ((firstToastRectangle.top + firstToastRectangle.height) - lastToastRectangle.top) + &#39;px&#39;;
                        } else {
                            $el.style.height = ((lastToastRectangle.top + lastToastRectangle.height) - firstToastRectangle.top) + &#39;px&#39;;
                        }
                    } else {
                        $el.style.height = firstToastRectangle.height + &#39;px&#39;;
                    }
                }
            }" @set-toasts-layout.window="
                layout=event.detail.layout;
                if(layout == &#39;expanded&#39;){
                    expanded=true;
                } else {
                    expanded=false;
                }
                stackToasts();
            " @toast-show.window="
                event.stopPropagation();
                if(event.detail.position){
                    position = event.detail.position;
                }
                toasts.unshift({
                    id: &#39;toast-&#39; + Math.random().toString(16).slice(2),
                    show: false,
                    message: event.detail.message,
                    description: event.detail.description,
                    type: event.detail.type,
                    html: event.detail.html
                });
            " @mouseenter="toastsHovered=true;" @mouseleave="toastsHovered=false" x-init="
                if(layout == &#39;expanded&#39;){
                    expanded = true;
                }
                stackToasts();
                $watch(&#39;toastsHovered&#39;, function(value){

                    if(layout == &#39;default&#39;){
                        if(position.includes(&#39;bottom&#39;)){
                            resetBottom();
                        } else {
                            resetTop();
                        }

                        if(value){
                            // calculate the new positions
                            expanded = true;
                            if(layout == &#39;default&#39;){
                                stackToasts();
                            }
                        } else {
                            if(layout == &#39;default&#39;){
                                expanded = false;
                                //setTimeout(function(){
                                stackToasts();
                            //}, 10);
                                setTimeout(function(){
                                    stackToasts();
                                }, 10)
                            }
                        }
                    }
                });
            " class="fixed block w-full group z-[99] sm:max-w-xs" :class="{ &#39;right-0 top-0 sm:mt-6 sm:mr-6&#39;: position==&#39;top-right&#39;, &#39;left-0 top-0 sm:mt-6 sm:ml-6&#39;: position==&#39;top-left&#39;, &#39;left-1/2 -translate-x-1/2 top-0 sm:mt-6&#39;: position==&#39;top-center&#39;, &#39;right-0 bottom-0 sm:mr-6 sm:mb-6&#39;: position==&#39;bottom-right&#39;, &#39;left-0 bottom-0 sm:ml-6 sm:mb-6&#39;: position==&#39;bottom-left&#39;, &#39;left-1/2 -translate-x-1/2 bottom-0 sm:mb-6&#39;: position==&#39;bottom-center&#39; }" x-cloak>
        <template x-for="(toast, index) in toasts" :key="toast.id">
          <li :id="toast.id" x-data="{
                        toastHovered: false
                    }" x-init="
                        
                        if(position.includes(&#39;bottom&#39;)){
                            $el.firstElementChild.classList.add(&#39;toast-bottom&#39;);
                            $el.firstElementChild.classList.add(&#39;opacity-0&#39;, &#39;translate-y-full&#39;);
                        } else {
                            $el.firstElementChild.classList.add(&#39;opacity-0&#39;, &#39;-translate-y-full&#39;);
                        }
                        setTimeout(function(){
                            
                            setTimeout(function(){
                                if(position.includes(&#39;bottom&#39;)){
                                    $el.firstElementChild.classList.remove(&#39;opacity-0&#39;, &#39;translate-y-full&#39;);
                                } else {
                                    $el.firstElementChild.classList.remove(&#39;opacity-0&#39;, &#39;-translate-y-full&#39;);
                                }
                                $el.firstElementChild.classList.add(&#39;opacity-100&#39;, &#39;translate-y-0&#39;);

                                setTimeout(function(){
                                    stackToasts();
                                }, 10);
                            }, 5);
                        }, 50);
        
                        setTimeout(function(){
                            setTimeout(function(){
                                $el.firstElementChild.classList.remove(&#39;opacity-100&#39;);
                                $el.firstElementChild.classList.add(&#39;opacity-0&#39;);
                                if(toasts.length == 1){
                                    $el.firstElementChild.classList.remove(&#39;translate-y-0&#39;);
                                    $el.firstElementChild.classList.add(&#39;-translate-y-full&#39;);
                                }
                                setTimeout(function(){
                                    deleteToastWithId(toast.id)
                                }, 300);
                            }, 5);
                        }, 9000); 
                    " @mouseover="toastHovered=true" @mouseout="toastHovered=false" class="absolute w-full duration-300 ease-out select-none sm:max-w-xs" :class="{ &#39;toast-no-description&#39;: !toast.description }">
            <span class="relative flex flex-col items-start shadow-[0_5px_15px_-3px_rgb(0_0_0_/_0.08)] w-full transition-all duration-300 ease-out border border-base-300 bg-base-300/90 sm:rounded-md sm:max-w-xs group p-4">
              <template x-if="!toast.html">
                <div class="relative">
                  <div class="flex items-center" :class="{ &#39;text-green-500&#39; : toast.type==&#39;success&#39;, &#39;text-blue-500&#39; : toast.type==&#39;info&#39;, &#39;text-orange-400&#39; : toast.type==&#39;warning&#39;, &#39;text-red-500&#39; : toast.type==&#39;danger&#39;, &#39;text-neutral-500&#39; : toast.type==&#39;default&#39; }">
                    <svg x-show="toast.type==&#39;success&#39;" class="w-[18px] h-[18px] mr-1.5 -ml-1" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                      <path fill-rule="evenodd" clip-rule="evenodd" d="M12 2C6.47715 2 2 6.47715 2 12C2 17.5228 6.47715 22 12 22C17.5228 22 22 17.5228 22 12C22 6.47715 17.5228 2 12 2ZM16.7744 9.63269C17.1238 9.20501 17.0604 8.57503 16.6327 8.22559C16.2051 7.87615 15.5751 7.93957 15.2256 8.36725L10.6321 13.9892L8.65936 12.2524C8.24484 11.8874 7.61295 11.9276 7.248 12.3421C6.88304 12.7566 6.92322 13.3885 7.33774 13.7535L9.31046 15.4903C10.1612 16.2393 11.4637 16.1324 12.1808 15.2547L16.7744 9.63269Z" fill="currentColor"></path>
                    </svg>
                    <svg x-show="toast.type==&#39;info&#39;" class="w-[18px] h-[18px] mr-1.5 -ml-1" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                      <path fill-rule="evenodd" clip-rule="evenodd" d="M12 2C6.47715 2 2 6.47715 2 12C2 17.5228 6.47715 22 12 22C17.5228 22 22 17.5228 22 12C22 6.47715 17.5228 2 12 2ZM12 9C12.5523 9 13 8.55228 13 8C13 7.44772 12.5523 7 12 7C11.4477 7 11 7.44772 11 8C11 8.55228 11.4477 9 12 9ZM13 12C13 11.4477 12.5523 11 12 11C11.4477 11 11 11.4477 11 12V16C11 16.5523 11.4477 17 12 17C12.5523 17 13 16.5523 13 16V12Z" fill="currentColor"></path>
                    </svg>
                    <svg x-show="toast.type==&#39;warning&#39;" class="w-[18px] h-[18px] mr-1.5 -ml-1" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                      <path fill-rule="evenodd" clip-rule="evenodd" d="M9.44829 4.46472C10.5836 2.51208 13.4105 2.51168 14.5464 4.46401L21.5988 16.5855C22.7423 18.5509 21.3145 21 19.05 21L4.94967 21C2.68547 21 1.25762 18.5516 2.4004 16.5862L9.44829 4.46472ZM11.9995 8C12.5518 8 12.9995 8.44772 12.9995 9V13C12.9995 13.5523 12.5518 14 11.9995 14C11.4473 14 10.9995 13.5523 10.9995 13V9C10.9995 8.44772 11.4473 8 11.9995 8ZM12.0009 15.99C11.4486 15.9892 11.0003 16.4363 10.9995 16.9886L10.9995 16.9986C10.9987 17.5509 11.4458 17.9992 11.9981 18C12.5504 18.0008 12.9987 17.5537 12.9995 17.0014L12.9995 16.9914C13.0003 16.4391 12.5532 15.9908 12.0009 15.99Z" fill="currentColor"></path>
                    </svg>
                    <svg x-show="toast.type==&#39;danger&#39;" class="w-[18px] h-[18px] mr-1.5 -ml-1" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                      <path fill-rule="evenodd" clip-rule="evenodd" d="M2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12ZM11.9996 7C12.5519 7 12.9996 7.44772 12.9996 8V12C12.9996 12.5523 12.5519 13 11.9996 13C11.4474 13 10.9996 12.5523 10.9996 12V8C10.9996 7.44772 11.4474 7 11.9996 7ZM12.001 14.99C11.4488 14.9892 11.0004 15.4363 10.9997 15.9886L10.9996 15.9986C10.9989 16.5509 11.446 16.9992 11.9982 17C12.5505 17.0008 12.9989 16.5537 12.9996 16.0014L12.9996 15.9914C13.0004 15.4391 12.5533 14.9908 12.001 14.99Z" fill="currentColor"></path>
                    </svg>
                    <p class="text-[13px] font-medium leading-none text-neutral-500" x-text="toast.message"></p>
                  </div>
                  <p x-show="toast.description" :class="{ &#39;pl-5&#39; : toast.type!=&#39;default&#39; }" class="mt-1.5 text-xs leading-none opacity-70" x-text="toast.description"></p>
                </div>
              </template>
              <span @click="burnToast(toast.id)" class="absolute right-0 p-1.5 mr-2.5 text-current duration-100 ease-in-out rounded-full opacity-0 cursor-pointer hover:bg-neutral-500" :class="{ &#39;top-1/2 -translate-y-1/2&#39; : !toast.description, &#39;top-0 mt-2.5&#39; : toast.description, &#39;opacity-100&#39; : toastHovered, &#39;opacity-0&#39; : !toastHovered }">
                <svg class="w-3 h-3" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
                  <path fill-rule="evenodd" d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z" clip-rule="evenodd"></path>
                </svg>
              </span>
            </span>
          </li>
        </template>
      </ul>
    </template>
    <script type="text/javascript">
      document.addEventListener("showToast", (e) => {
      const level = e.detail.level;
      const message = e.detail.message;
      window.toast(message, {type: level});
      });
    </script>
  </body>
</html>
---
